rust/foxglove_data_loader/examples/ndjson.rs

```diff
@@ -0,0 +1,374 @@
+struct NDJsonIterator {
+    rows: Rc<RefCell<Vec<Row>>>,
+    index: RefCell<usize>,
+    start: u64,
+    end: u64,
+    channels: BTreeSet<u16>,
+}
+
+impl NDJsonIterator {
+    fn open(rows: Rc<RefCell<Vec<Row>>>, args: &MessageIteratorArgs) -> Self {
+        Self {
+            rows: rows.clone(),
+            index: 0.into(),
+            start: args.start_time.unwrap_or(0),
+            end: args.end_time.unwrap_or(u64::MAX),
+            channels: args.channels.iter().copied().collect(),
+        }
+    }
+}
+
+impl MessageIterator for NDJsonIterator {
+    type Error = anyhow::Error;
+
+    fn next(&self) -> Option<Result<Message, Self::Error>> {
+        loop {
+            let index = *self.index.borrow();
+            self.index.replace(index + 1);
+            let rows = self.rows.borrow();
+            let row = rows.get(index);
+            if let Some(time) = row.and_then(|r| r.get_time().map(seconds_to_nanos)) {
+                if time < self.start {
+                    continue;
+                }
+                if time > self.end {
+                    return None;
+                }
+            };
+            match row {
+                None => return None,
+                Some(Row::Info(_)) => {}
+                Some(Row::Sphere(sphere)) => {
+                    if self.channels.contains(&SPHERE_CHANNEL_ID) {
+                        return Some(Ok(sphere.into()));
+                    }
+                }
+                Some(Row::Temperature(temperature)) => {
+                    if self.channels.contains(&TEMPERATURE_CHANNEL_ID) {
+                        return Some(Ok(temperature.into()));
+                    }
+                }
+            };
+        }
+    }
+}
+
+// floating point time in seconds to u64 nanoseconds
+fn seconds_to_nanos(time: f64) -> u64 {
+    (time * 1.0e9) as u64
+}
+
+fn get_timestamp(time: f64) -> Timestamp {
+    let time_nanos = seconds_to_nanos(time);
+    Timestamp::new(
+        (time_nanos / 1_000_000_000) as u32,
+        (time_nanos % 1_000_000_000) as u32,
+    )
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(tag = "type")]
+enum Row {
+    #[serde(rename = "info")]
+    Info(Info),
+    #[serde(rename = "sphere")]
+    Sphere(Sphere),
+    #[serde(rename = "temperature")]
+    Temperature(Temperature),
+}
+
+impl Row {
+    fn get_time(&self) -> Option<f64> {
+        match self {
+            Row::Info(_) => None,
+            Row::Sphere(sphere) => Some(sphere.time),
+            Row::Temperature(temperature) => Some(temperature.time),
+        }
+    }
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct Info {
+    start: f64,
+    end: f64,
+    sphere_count: u64,
+    temperature_count: u64,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct Sphere {
+    id: String,
+    time: f64,
+    x: f64,
+    y: f64,
+    z: f64,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
```

Comment:
You don't need to define the schema below If you derive `foxglove::Encode` here. This generates a schema for you and handles encoding into protobuf. See also https://docs.rs/foxglove/latest/foxglove/#custom-data

Example where this is used: 
https://github.com/foxglove/mp4-data-loader/blob/49c27d5707cb5bb6eb5f4b152d8886e25e00f99b/data-loader/src/lib.rs#L16-L17 (see that same file for how to get the schema and encode messages)